src/aspire/orientation/__init__.py

```diff
@@ -1,38 +1,556 @@
+                dx = -max_shift + sidx2 * shift_step
+
+            idx = np.arange(4 * shift_eq_idx, 4 * shift_eq_idx + 4)
+            shift_alpha = c_ij * d_theta
+            shift_beta = c_ji * d_theta
+            shift_i[idx] = shift_eq_idx
+            shift_j[idx] = [2 * i, 2 * i + 1, 2 * j, 2 * j + 1]
+            shift_b[shift_eq_idx] = dx
+
+            # check with compare cl
+            if not is_pf_j_flipped:
+                shift_eq[idx] = [np.sin(shift_alpha), np.cos(shift_alpha),
+                                 -np.sin(shift_beta), -np.cos(shift_beta)]
+            else:
+                shift_beta -= np.pi
+                shift_eq[idx] = [-np.sin(shift_alpha), -np.cos(shift_alpha),
+                                 -np.sin(shift_beta), -np.cos(shift_beta)]
+
+        t = 4 * n_equations
+        shift_eq[t: t + n_equations] = shift_b
+        shift_i[t: t + n_equations] = np.arange(n_equations)
+        shift_j[t: t + n_equations] = 2 * n_img
+        mask = np.where(shift_eq != 0)[0]
+        shift_eq = shift_eq[mask]
+        shift_i = shift_i[mask]
+        shift_j = shift_j[mask]
+        shift_equations = sps.csr_matrix((shift_eq, (shift_i, shift_j)),
+                                         shape=(n_equations, 2 * n_img + 1))
+
+        est_shifts = sps.linalg.lsqr(shift_equations[:, :-1], shift_b)[0]
+        est_shifts = est_shifts.reshape((2, n_img), order='F')
+
+        return est_shifts, shift_equations
+
+    def _estimate_num_shift_equations(self, n_img, memory_factor):
         """
-        find the common lines by voting method
+        Estimate total number of shift equations in images
+
+        The function computes total number of shift equations based on
+        number of images and preselected memory factor.
+        :param n_img:  The total number of input images
+        :param memory_factor: If there are N projections, then the system of
+            equations solved for the shifts is of size 2N x N(N-1)/2 (2N
+            unknowns and N(N-1)/2 equations). This may be too big if N is
+            large. If `memory_factor` between 0 and 1, then it is the
+            fraction of equation to retain. That is, the system of
+            equations solved will be of size 2N x N*(N-1)/2*`memory_factor`.
+            If `memory_factor` is larger than 100, then the number of
+            equations is estimated in such a way that the memory used by the equations
+            is roughly `memory_factor megabytes`. Default is 10000 (use all equations).
+            The code will generate an error if `memory_factor` is between 1 and 100.
+        :return: Estimated number of shift equations
         """
-        pass
+        # Number of equations that will be used to estimation the shifts
+        n_equations_total = int(np.ceil(n_img * (n_img - 1) / 2))
+        # Estimated memory requirements for the full system of equation.
+        # This ignores the sparsity of the system, since backslash seems to
```

Comment:
This is only an estimated number for the full system ignores the sparsity. It will be reduced based on the `memory_factor.`

Reply:
Ok. But why are we ignoring the sparsity of the system? We don't have `n_img * (n_img - 1) / 2` comparisons since `n_check < n_img`.

Reply:
Changed  `n_img * (n_img - 1) / 2` to `n_img * (n_check - 1) / 2` in new commit.

Reply:
Isn't it `n_img * n_check`?

Reply:
We set `n_check = n_img` if it is not specific so that the exact number of shift equations is `n_img * (n_img - 1) / 2`. This is for consistent purpose only. In the really calculation this value is setting an upbound and only an estimated number of equations are used.

Reply:
It depends on what kind of computer used. I set to 4GB so that most of laptops and desktops should be fine now.

Reply:
Okay that's a start. Thanks